File: 2020/02/rs/src/main.rs
use std::env;
use std::fs;
use std::time::{Instant};
use regex::Regex;

type Line = (usize, usize, char, String);

macro_rules! count_valid {
    ($lines: expr, $filter: expr) => {
        $lines.iter().filter($filter).count()
    };
}

fn solve(lines: &Vec<Line>) -> (usize,usize) {
    (
        count_valid!(lines, |line| {
            let (minimum, maximum, letter, password) = line;
            let occurence_count = password.chars().filter(|c| c == letter).count();
            occurence_count >= *minimum && occurence_count <= *maximum
        }),
        count_valid!(lines, |line| {
            let (first, second, letter, password) = line;
            (password.chars().nth(first - 1).unwrap() == *letter) 
            ^ 
            (password.chars().nth(second - 1).unwrap() == *letter)
        })
    )
}

fn get_input(file_path: &String) -> Vec<Line> {
    let re = Regex::new(r"^(\d+)-(\d+)\s([a-z]):\s(.*)$").unwrap();
    fs::read_to_string(file_path).expect("Error reading input file!")
        .lines()
        .map(|line| re.captures(line).and_then(|caps| 
            Some((
                caps.get(1)
                    .expect("Group not found in capture")
                    .as_str()
                    .parse()
                    .expect("Unable to parse number"),
                caps.get(2)
                    .expect("Group not found in capture")
                    .as_str()
                    .parse()
                    .expect("Unable to parse number"),
                caps.get(3)
                    .expect("Group not found in capture")
                    .as_str()
                    .chars()
                    .next()
                    .expect("Unable to retrieve first char of string"),
                String::from(caps.get(4)
                    .expect("Group not found in capture")
                    .as_str()
                )
            ))
        ).expect("Bad format"))
        .collect()
}

fn main() {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        panic!("Please, add input file path as parameter");
    }
    let now = Instant::now();
    let (part1_result, part2_result) = solve(&get_input(&args[1]));
    println!("P1: {}", part1_result);
    println!("P1: {}", part2_result);
    println!();
    println!("Time: {:.7}", now.elapsed().as_secs_f32());
}

File: 2020/02/c/main.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/time.h>
#include <regex.h>
#include <string.h>

#define MAX_COUNT 1024

typedef struct
{
    int first;
    int second;
    char letter;
    char *password;
} Line;

typedef struct
{
    Line *lines;
    int count;
} Input;

typedef struct
{
    int part1;
    int part2;
} Results;

int countValid(Input input, int (*validationFunc)(Line *))
{
    int count = 0;
    while (input.count--)
        count += validationFunc(input.lines++);
    return count;
}

int isLineValid1(Line *line)
{
    char *password = line->password;
    int occurenceCount = 0;
    char c;
    while ((c = *password++))
        occurenceCount += c == line->letter;
    return occurenceCount >= line->first && occurenceCount <= line->second;
}

int isLineValid2(Line *line)
{
    return (line->password[line->first - 1] == line->letter) ^ (line->password[line->second - 1] == line->letter);
}

Results solve(Input input)
{
    return (Results){countValid(input, isLineValid1), countValid(input, isLineValid2)};
}

Input getInput(char *filePath)
{
    FILE *file;
    if ((file = fopen(filePath, "r")) == NULL)
    {
        perror("Error reading input file!\n");
        exit(1);
    }
    regex_t regexCompiled;
    regmatch_t groupArray[MAX_COUNT];
    if (regcomp(&regexCompiled, "^([0-9]+)-([0-9]+) ([a-z]): (.*)$", REG_EXTENDED))
    {
        perror("Error compiling regex.");
        exit(1);
    }
    Line *lines = calloc(MAX_COUNT, sizeof(Line));
    int count = 0, group;
    size_t lineLength;
    char *line = NULL, *cursor = NULL;
    while (getline(&line, &lineLength, file) != EOF)
    {
        cursor = line;
        int first, second;
        char letter;
        char *password = malloc(32);
        while (!regexec(&regexCompiled, cursor, MAX_COUNT, groupArray, 0))
        {
            for (group = 0; group <= 4 && groupArray[group].rm_so != -1; group++)
            {
                char cursorCopy[strlen(cursor) + 1];
                strcpy(cursorCopy, cursor);
                cursorCopy[groupArray[group].rm_eo] = 0;
                switch (group)
                {
                case 1:
                    first = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 2:
                    second = atoi(cursorCopy + groupArray[group].rm_so);
                    break;
                case 3:
                    letter = cursorCopy[groupArray[group].rm_so];
                    break;
                case 4:
                    strcpy(password, cursorCopy + groupArray[group].rm_so);
                    break;
                }
            }
            cursor += groupArray[0].rm_eo;
        }
        lines[count++] = (Line){first, second, letter, password};
    }
    regfree(&regexCompiled);
    fclose(file);
    return (Input){lines, count};
}

void freeInput(Input input)
{
    Line *lines = input.lines;
    while (input.count--)
        free((lines++)->password);
    free(input.lines);
}

int main(int argc, char **argv)
{
    if (argc != 2)
    {
        perror("Please, add input file path as parameter");
        exit(1);
    }
    struct timeval starts, ends;
    gettimeofday(&starts, NULL);
    Input input = getInput(argv[1]);
    Results results = solve(input);
    gettimeofday(&ends, NULL);
    freeInput(input);
    printf("P1: %d\n", results.part1);
    printf("P2: %d\n\n", results.part2);
    printf("Time: %.7f\n", (double)((ends.tv_sec - starts.tv_sec) * 1000000 + ends.tv_usec - starts.tv_usec) / 1000000);
    return 0;
}
